src/unmangle.rs

```diff
@@ -0,0 +1,102 @@
+//! Path unmangling functions.
+//!
+//! Paths included in several files under `/proc` are mangled, i.e., some
+//! characters are replaced by the corresponding octal escape sequence `\nnn`.
+//!
+//! For example, the following files:
+//!
+//! - `/proc/[pid]/maps`
+//! - `/proc/[pid]/smaps`
+//! - `/proc/[pid]/numa_maps`
+//! - `/proc/swaps`
+//!
+//! contain mangled paths.
+//!
+//! This module provides the [`unmangled_path`] function to reverse the
+//! mangling (decoding the escape sequences).
+//!
+//! Note that, unless `\` is included in the set of the escaped characters
+//! (which is *not* the case in any of the previous files), the mangling is
+//! actually non-reversible (i.e., the demangling is ambiguous).
+//!
+//! See `mangle_path` in `Linux/fs/seq_file.c` for details on the mangling
+//! algorithm.
+//!
+//! [`unmangled_path`]: fn.unmangled_path.html
+
+use std::str;
+use std::num::ParseIntError;
+
+/// Converts a bytes slice in a given base to an integer.
+fn u8_from_bytes_radix(bytes: &[u8], radix: u32) -> Result<u8, ParseIntError> {
+    let s = unsafe { str::from_utf8_unchecked(bytes) };
+    u8::from_str_radix(s, radix)
+}
+
+/// Returns an iterator that yields the unmangled representation of `path` as
+/// bytes.
+///
+/// This struct is used by the [`unmangled_path`] function. See its
+/// documentation for more.
+///
+/// [`unmangled_path`]: fn.unmangled_path.html
+struct UnmangledPath<'a> {
+    /// Slice of the source path to be unmangled.
+    path: &'a [u8],
+    /// Escaped chars that should be decoded (e.g., b"\n ").
+    escaped: &'a [u8],
+}
+
+impl<'a> Iterator for UnmangledPath<'a> {
```

Comment:
The iterator here is obfuscating the logic, could you just fold it into `unmangled_path` as a for loop?

Reply:
Sure! The code does indeed look more straightforward without the iterator. Thanks for the feedback :-)